src/tribler/core/tests/test_search_utils.py

```diff
@@ -15,3 +21,129 @@ def test_filter_keywords():
+
+    # The more non-matching words are in the title, the worse is rank
+
+    r9 = torrent_rank(query, 'Big Buck A Bunny')  # 0.75061099
+    r10 = torrent_rank(query, 'Big Buck A B Bunny')  # 0.699335863
+    r11 = torrent_rank(query, 'Big Buck A B C Bunny')  # 0.6546181
+
+    assert r11 < r10 < r9 < r1
+
+    # Non-matching words close to the beginning of the title give a bigger penalty
+
+    r12 = torrent_rank(query, 'Big A Buck Bunny')  # 0.742420681
+    r13 = torrent_rank(query, 'Big A B Buck Bunny')  # 0.6852494577
+    r14 = torrent_rank(query, 'Big A B C Buck Bunny')  # 0.636253776
+
+    assert r14 < r13 < r12 < r1
+
+    r15 = torrent_rank(query, 'A Big Buck Bunny')  # 0.73125
+    r16 = torrent_rank(query, 'A B Big Buck Bunny')  # 0.66645569
+    r17 = torrent_rank(query, 'A B C Big Buck Bunny')  # 0.6122093
+
+    assert r17 < r16 < r15 < r1
+    assert r15 < r12 and r16 < r13 and r17 < r14
+
+    # Wrong order of words in the title imposes a penalty to the rank
+
+    r18 = torrent_rank(query, 'Big Bunny Buck')  # 0.7476923
+
+    assert r18 < r1
+
+    # Missed query words imposes a really big penalty
+
+    r19 = torrent_rank(query, 'Big Buck')  # 0.4725
+
+    assert r19 < 0.5
+
+    # The close the missed words to the beginning of the query, the worse
+
+    r20 = torrent_rank(query, 'Big Bunny')  # 0.441818181
+    r21 = torrent_rank(query, 'Buck Bunny')  # 0.405
+
+    assert r21 < r20 < r19
+
+    # The more seeders is still better
+
+    r22 = torrent_rank(query, 'Buck Bunny', seeders=10, freshness=5 * DAY)  # 0.44805194
+    r23 = torrent_rank(query, 'Buck Bunny', seeders=100, freshness=5 * DAY)  # 0.46821428
+    r24 = torrent_rank(query, 'Buck Bunny', seeders=1000, freshness=5 * DAY)  # 0.4883766
+
+    assert r21 < r22 < r23 < r24
+
+    # The more days from the check the less relevant the number of seeders is
+
+    r25 = torrent_rank(query, 'Buck Bunny', seeders=1000, freshness=10 * DAY)  # 0.48306818
+    r26 = torrent_rank(query, 'Buck Bunny', seeders=1000, freshness=20 * DAY)  # 0.47563636
+
+    assert r26 < r25 < r24
+
+    # The exact match has a good rank
+    r27 = torrent_rank('Sintel', 'Sintel')  # 0.81
+    assert r27 > 0.8
+
+    # Non-matching words at the end of the title give slightly worse results
+    r28 = torrent_rank('Sintel', 'Sintel Part II')  # 0.79553571
+
+    # Non-matching words at the beginning of the title are much worse
+    r29 = torrent_rank('Sintel', 'Part of Sintel')  # 0.664925373
+
+    # Too many non-matching words give a bigger penalty
+    r30 = torrent_rank('Sintel', 'the.script.from.the.movie.Sintel.pdf')  # 0.52105263
+
+    assert r30 < r29 < r28 < r27
+
+    # Some more examples
+
+    r31 = torrent_rank("Internet's Own Boy", "Internet's Own Boy")  # 0.81
+    r32 = torrent_rank("Internet's Own Boy", "Internet's very Own Boy")  # 0.75099337
+    r33 = torrent_rank("Internet's Own Boy", "Internet's very special Boy person")  # 0.4353166986
+
+    assert r33 < r32 < r31
```

Comment:
I've played with my own search implementation and found that the current test style is not convenient for bug hunting. 

For example, if `torrent_rank` is broken and the assertion `   assert r8 < r7 < r6 < r5 < r1` raises an error, then it is impossible (really hardly possible) to find out which exact`torrent_rank` invocation failed.

You will see in the test the following output:
```python
        # The closer to the start of the string non-matching words are placed in the title, the worse is rank
        r5 = torrent_rank(query, 'Big Buck Bunny II')  # 0.80381679
        r6 = torrent_rank(query, 'Big Buck Brown Bunny')  # 0.75061099
        r7 = torrent_rank(query, 'Big Bad Buck Bunny')  # 0.74242068
        r8 = torrent_rank(query, 'Boring Big Buck Bunny')  # 0.73125
    
>       assert r8 < r7 < r6 < r5 < r1
E       assert 0.72 < 0.670344827586207
``` 
It is really hard to guess which of the four invocations failed. 

To improve the test I suggest restructuring this test into the following:
```suggestion
def test_torrent_rank():
    query = 'Big Buck Bunny'

    # The exact match ranked as pretty high
    full_match = torrent_rank(query, 'Big Buck Bunny')  # 0.81
    assert full_match > 0.8

    # Seeders are good for the rank
    # The more seeders the better
    # The fewer days have passed since the creation of the torrent, the higher its rank
    assert torrent_rank(query, 'Big Buck Bunny', seeders=1000, freshness=1 * DAY) > \
           torrent_rank(query, 'Big Buck Bunny', seeders=1000, freshness=100 * DAY) > \
           torrent_rank(query, 'Big Buck Bunny', seeders=100, freshness=100 * DAY) > \
           full_match

    # If a title contains non-matching words missed in the query string it is not as good as the exact match
    # The closer to the start of the string non-matching words are placed in the title, the worse is rank
    assert full_match > \
           torrent_rank(query, 'Big Buck Bunny II') > \
           torrent_rank(query, 'Big Buck Brown Bunny') > \
           torrent_rank(query, 'Big Bad Buck Bunny') > \
           torrent_rank(query, 'Boring Big Buck Bunny')

    # The more non-matching words are in the title, the worse is rank
    assert full_match > \
           torrent_rank(query, 'Big Buck A Bunny') > \
           torrent_rank(query, 'Big Buck A B Bunny') > \
           torrent_rank(query, 'Big Buck A B C Bunny')

    # Non-matching words close to the beginning of the title give a bigger penalty
    assert full_match > \
           torrent_rank(query, 'Big A Buck Bunny') > \
           torrent_rank(query, 'Big A B Buck Bunny') > \
           torrent_rank(query, 'Big A B C Buck Bunny')

    assert full_match > \
           torrent_rank(query, 'A Big Buck Bunny') > \
           torrent_rank(query, 'A B Big Buck Bunny') > \
           torrent_rank(query, 'A B C Big Buck Bunny')

    assert torrent_rank(query, 'Big A Buck Bunny') > \
           torrent_rank(query, 'A Big Buck Bunny')
    assert torrent_rank(query, 'Big A B Buck Bunny') > \
           torrent_rank(query, 'A B Big Buck Bunny')
    assert torrent_rank(query, 'Big A B C Buck Bunny') > \
           torrent_rank(query, 'A B C Big Buck Bunny')

    # Wrong order of words in the title imposes a penalty to the rank
    assert full_match > \
           torrent_rank(query, 'Big Bunny Buck')

    # Missed query words imposes a really big penalty
    assert torrent_rank(query, 'Big Buck') < 0.5

    # The close the missed words to the beginning of the query, the worse
    assert torrent_rank(query, 'Big Buck') > \
           torrent_rank(query, 'Big Bunny') > \
           torrent_rank(query, 'Buck Bunny')

    # The more seeders is still better
    assert torrent_rank(query, 'Buck Bunny', seeders=1000, freshness=5 * DAY) > \
           torrent_rank(query, 'Buck Bunny', seeders=100, freshness=5 * DAY) > \
           torrent_rank(query, 'Buck Bunny', seeders=10, freshness=5 * DAY) > \
           torrent_rank(query, 'Buck Bunny')

    # The more days from the check the less relevant the number of seeders is
    assert torrent_rank(query, 'Buck Bunny', freshness=5 * DAY) > \
           torrent_rank(query, 'Buck Bunny', freshness=10 * DAY) > \
           torrent_rank(query, 'Buck Bunny', freshness=20 * DAY)

    # The exact match has a good rank
    assert torrent_rank('Sintel', 'Sintel') > 0.8

    # Non-matching words at the end of the title give slightly worse results
    # Non-matching words at the beginning of the title are much worse
    # Too many non-matching words give a bigger penalty
    assert torrent_rank('Sintel', 'Sintel') > \
           torrent_rank('Sintel', 'Sintel Part II') > \
           torrent_rank('Sintel', 'Part of Sintel') > \
           torrent_rank('Sintel', 'the.script.from.the.movie.Sintel.pdf')

    # Some more examples
    assert torrent_rank("Internet's Own Boy", "Internet's Own Boy") > \
           torrent_rank("Internet's Own Boy", "Internet's very Own Boy") > \
           torrent_rank("Internet's Own Boy", "Internet's very special Boy person")
```

After this change, the test output will be the following:

```python
>       assert full_match > \
               torrent_rank(query, 'Big Buck Bunny II') > \
               torrent_rank(query, 'Big Buck Brown Bunny') > \
               torrent_rank(query, 'Big Bad Buck Bunny') > \
               torrent_rank(query, 'Boring Big Buck Bunny')
E       AssertionError: assert 0.670344827586207 > 0.72
E        +  where 0.670344827586207 = torrent_rank('Big Buck Bunny', 'Big Buck Brown Bunny')
E        +  and   0.72 = torrent_rank('Big Buck Bunny', 'Big Bad Buck Bunny')
```

Reply:
Thanks, I rewrited the tests according to your suggestion

Reply:
🙏